pkg/option/option.go

```diff
@@ -225,75 +260,77 @@ func (opt *Options) adjust() {
+		}
+	}
 	switch opt.ClusterRole {
 	case "reader":
 		if opt.ForceNewCluster {
 			return fmt.Errorf("reader got force-new-cluster")
 		}
-
-		if len(opt.ClusterJoinURLs) == 0 {
+		if !opt.UseInitialCluster() && len(opt.ClusterJoinURLs) == 0 {
 			return fmt.Errorf("reader got empty cluster-join-urls")
 		}
-
-		for _, urlText := range opt.ClusterJoinURLs {
-			_, err := url.Parse(urlText)
-			if err != nil {
-				return fmt.Errorf("invalid cluster-join-urls: %v", err)
-			}
-		}
-
 	case "writer":
-		if len(opt.ClusterListenClientURLs) == 0 {
-			return fmt.Errorf("empty cluster-listen-client-urls")
-		}
-		if len(opt.ClusterListenPeerURLs) == 0 {
-			return fmt.Errorf("empty cluster-listen-peer-urls")
-		}
-		if len(opt.ClusterAdvertiseClientURLs) == 0 {
-			return fmt.Errorf("empty cluster-advertise-client-urls")
-		}
-		if len(opt.ClusterInitialAdvertisePeerURLs) == 0 {
-			return fmt.Errorf("empty cluster-initial-advertise-peer-urls")
+		argumentsToValidate := map[string][]string{
+			"cluster-listen-client-urls":          opt.ClusterListenClientURLs,
+			"cluster-listen-peer-urls":            opt.ClusterListenPeerURLs,
+			"cluster-advertise-client-urls":       opt.ClusterAdvertiseClientURLs,
+			"cluster-initial-advertise-peer-urls": opt.ClusterInitialAdvertisePeerURLs,
 		}
 
-		for _, clientURL := range opt.ClusterListenClientURLs {
-			_, err := url.Parse(clientURL)
-			if err != nil {
-				return fmt.Errorf("invalid cluster-listen-client-urls: %s: %v", clientURL, err)
+		if opt.UseInitialCluster() {
+			argumentsToValidate = map[string][]string{
+				"listen-client-urls":          opt.Cluster.ListenClientURLs,
+				"listen-peer-urls":            opt.Cluster.ListenPeerURLs,
+				"advertise-client-urls":       opt.Cluster.AdvertiseClientURLs,
+				"initial-advertise-peer-urls": opt.Cluster.InitialAdvertisePeerURLs,
 			}
-		}
-		for _, peerURL := range opt.ClusterListenPeerURLs {
-			_, err := url.Parse(peerURL)
-			if err != nil {
-				return fmt.Errorf("invalid cluster-listen-peer-urls: %s: %v", peerURL, err)
+			initialClusterUrls := make([]string, 0, len(opt.Cluster.InitialCluster))
+			for _, value := range opt.Cluster.InitialCluster {
+				initialClusterUrls = append(initialClusterUrls, value)
 			}
-		}
-		for _, clientURL := range opt.ClusterAdvertiseClientURLs {
-			_, err := url.Parse(clientURL)
-			if err != nil {
-				return fmt.Errorf("invalid cluster-advertise-client-urls: %s: %v", clientURL, err)
+			if _, err := ParseURLs(initialClusterUrls); err != nil {
+				return fmt.Errorf("invalid initial-cluster: %v", err)
 			}
-		}
-		for _, peerURL := range opt.ClusterInitialAdvertisePeerURLs {
-			_, err := url.Parse(peerURL)
-			if err != nil {
-				return fmt.Errorf("invalid cluster-initial-advertise-peer-urls: %s: %v", peerURL, err)
+			if len(opt.ClusterJoinURLs) > 0 {
+				err := strings.Join(strings.Fields(`
+					%s and %s are both defined.
+					Please provide only one of them. %s is the recommended way.`), " ")
```

Comment:
please keep this as a string, it takes me some time to realize the purpose (maybe we can refactor this function to avoid the long code line).
and we need to find another variable name because we use `err` as an `error` in all other cases.